### DISCUSSION TITLE
Automatically import a missing import (python)

### DISCUSSION THREAD

**Discussion by bakert** [zed-industries/zed#13281] [Feature Requests]
Loving Zed and of course its speed. I'm trying to get everything set up as effectively as possible. I would love to be able to add missing imports in Python code from the line where the import is first used.

<img width="533" alt="image" src="https://github.com/zed-industries/zed/assets/1541/bd4b8def-e656-4274-a279-f088e7c34490">

I see from https://github.com/zed-industries/zed/discussions/13115 that it's possible to add a missing import when working TypeScript. Does such a facility exist for Python?

![image](https://github.com/zed-industries/zed/assets/1541/f2a2b869-93f6-4dd1-9e7d-866806aa32a9)

I am using Pyright which magically set itself up for me when I started editing python files in Zed (nice!) Might I expect it to appear under Code Actions? (It does not.) Or in the left gutter with a lightning symbol like in the TypeScript example above? (What is the difference between these two things?)

Is this the responsibility of the LSP, or of Zed?

Thanks!

**Comment by linde12** [comment]
No code actions seem to be supported by pyright in zed, makes it kind of unusable for python :(

**Comment by Cernunnos** [comment]
When you type `path` you get the option to automatically import the function -> `from os import path`.
![zed](https://github.com/user-attachments/assets/927d4525-f877-49ef-80d2-8bc8ede01bf0)

For `exit` you get the option `from os import _exit`

**Reply by bakert** [reply]
> When you type `path` you get the option to automatically import the function -> `from os import path`. ![zed](https://private-user-images.githubusercontent.com/905002/349644036-927d4525-f877-49ef-80d2-8bc8ede01bf0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjE3NjcxNDMsIm5iZiI6MTcyMTc2Njg0MywicGF0aCI6Ii85MDUwMDIvMzQ5NjQ0MDM2LTkyN2Q0NTI1LWY4NzctNDllZi04MGQyLThiYzhlZGUwMWJmMC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNzIzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDcyM1QyMDM0MDNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hZWUzN2NiY2NjNTUxY2JkMTA3YjZkOWQ1MjQyMDczMThlODJhMTFkZjU2NDMzZWZjMWYyMzFiZjQ0ODcxMDRjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.hFSCYZFMjxigD16LeRQbeZG9nYTU09lMSvClU2Px92o)
> 
> For `exit` you get the option `from os import _exit`

That's interesting. Did you have to do anything to get to that state?

This is my experience typing `path` into an empty (but already saved as test.py) file:

<img width="402" alt="image" src="https://github.com/user-attachments/assets/7f076923-834c-40f4-863a-76e70d63e88d">

Is there something I need to configure?

**Reply by Cernunnos** [reply]
Hmm, I just tried it out again. What I noticed is that it seems to need a venv for this (must then be specified in the pyrightconfig.json).
Without venv, I get the same result in your screenshot.

**Reply by linde12** [reply]
This work, but there are still no code actions appearing when doing ToggleCodeActions. Coming from a coc.nvim workflow where all i have to do on an unimported symbol is "<leader>af" its a real setback :/

**Comment by albertwh1te** [comment]
I encountered the same issue. @Cernunnos's method works, but it would be ideal if it could generate imports like Visual Studio Code does.

![image](https://github.com/user-attachments/assets/4deff450-90c9-4fdc-a22d-a8797f021c8c)

**Comment by johnnybakucz** [comment]
What I found is that auto import only works if the symbol was typed. In case a completion is accepted from the assistant, the import is not handled. You have to go through all of the missing symbols and delete them and type them by hand, which is tedious and quite frustrating.

Today I set up zed as my main editor (switched from vs code), but this feature is really missing.

**Comment by simjak** [comment]
still no auto import?

**Reply by linde12** [reply]
I just checked on latest linux preview, no code actions suggested when using ToggleCodeActions.

**Comment by RhetTbull** [comment]
I am getting auto imports with python and I *hate*, *hate*, *hate* this feature. To me, it makes Zed unusable. I didn't turn this on and there's nothing in my settings.json that should have enabled this. It started happening recently on macOS and I can't figure out how to turn this off. I'm not even getting an option in intellisense to add the import (which seems broken recently...separate issue, see #22266). I cannot keep using zed if it throws random import statements into my code. Anyone know how to turn this off?

**Reply by failable** [reply]
You can turn off it by adding 

```json
"lsp": {
    "pyright": {
      "settings": {
        "python.analysis": {
          "autoImportCompletions": false
        }
      }
    }
  },
```

**Reply by JairajSustained** [reply]
Thank you so much

**Reply by RhetTbull** [reply]
Thank you!

**Reply by jabr** [reply]
This doesn't seem to work any more. I'm getting random, wrong imports added to my code again from when this (or some similar) auto import logic goes off the rails yet again...

**Reply by jabr** [reply]
This seems to help so far:
```
  "lsp": {
    "pyright": {
      "settings": {
        "pyright": {
          "disableLanguageServices": true
        }
      }
    }
  }
```

**Reply by NinZeige** [reply]
For basedpyright users, this would help. Refer: https://docs.basedpyright.com/v1.37.0/configuration/language-server-settings/ 

```json
  "lsp": {
    "basedpyright": {
      "settings": {
        "basedpyright.analysis": {
          "autoImportCompletions": false,
        },
      },
    },
  },
  ```

**Comment by 2zqa** [comment]
Using [basedpyright](https://github.com/m1guer/basedpyright-zed) solves this issue. Apparently this is a [pylance feature](https://github.com/microsoft/pyright/discussions/7099#discussioncomment-8215891), not a pyright feature. 

![Schermafdruk van 2025-02-07 15-30-32-obfuscated](https://github.com/user-attachments/assets/1a9b9c74-c60a-4bf7-b7b8-56a441fd8b95)

**Reply by KhoaLee** [reply]
Hi, I have installed and configure basedpyright but it seems like I still not get auto import suggestion like yours. Could you share for lsp and languages setup in settings.json of zed, also the pyproject.toml file for basedpyright?

**Reply by failable** [reply]
Those are code actions. Use `editor: toggle code actions` command to pop it up.

**Reply by 2zqa** [reply]
The default shortcut for that is <kbd>Ctrl</kbd> + <kbd>.</kbd>. A lot of IDEs use this shortcut for their code actions shortcuts, like Eclipse or even BlueJ.

**Reply by KhoaLee** [reply]
Sorry for my unclear question. I know that is code actions and the short cut to toggle it. However I have solve this problem by having to install basedpyright via npm to my local system. Do you need to install basedpyright via npm to local system? I did installed basedpyright extension in zed and proper configure language service. Unfortunately I have to do the install step to make it work. Thanks all

**Reply by 2zqa** [reply]
> Do you need to install basedpyright via npm to local system?

You need to have installed basedpyright in any way. I have installed it through pip.

Indeed, basedpyright needs to be installed as well as the basedpyright Zed extension for things to work.

**Comment by JosephTLyons** [comment]
Thanks for reaching out. We want to better support our community, and the best way for us to do that is through our official Discord. The team is active there and happy to help.

We’re closing the “Help and General Q&A” category on GitHub Discussions and directing all support questions to Discord going forward.

https://zed.dev/community-links
